pkg/daemon/daemon.go

```diff
@@ -1978,6 +2118,30 @@ func (dn *Daemon) completeUpdate(desiredConfigName string) error {
 	return nil
 }
 
+func (dn *Daemon) triggerUpdate(currentConfig, desiredConfig *mcfgv1.MachineConfig, currentImage, desiredImage string) error {
```

Comment:
I wanted to put an idea I had here for posterity:

I think some of the confusion and complexity around what the MCD does and when it happens comes from the the practice of passing many things around using positional function arguments. As we pass them around, their meaning changes depending on how far down the callstack you go.

For example, if you look at the args for `dn.triggerUpdateWithMachineConfig()` and `dn.update()`, respectively, you'll see that the meaning changes from one function to the next. In `dn.triggerUpdateWithMachineConfig()`, the meaning of `currentConfig` and `desiredConfig` remains intact. However, in `dn.update()`, their meaning changes from `currentConfig` / `desiredConfig` to `oldConfig` / `newConfig`. Sure, the values are the same, but the meaning behind them changes. Additionally, given that the function signatures look like this: `triggerUpdateWithMachineConfig(currentConfig, desiredConfig *mcfgv1.MachineConfig, ...)`, it becomes easy to mistakenly pass the wrong value in, causing bugs and further confusion.

Here is a proposed solution that involves the addition of a few new structs and transformation methods which make this transition explicit instead of implicit:
```golang
// This struct represents the updates we're about to apply using the old / new nomenclature.
type updateToApply struct {
	oldConfig *mcfgv1.MachineConfig
	newConfig *mcfgv1.MachineConfig
	oldImage string
	newImage string
}


// This generates the onDiskConfig data that gets written to disk.
func (u *updateToApply) toOnDiskConfig() *onDiskConfig {
	return &onDiskConfig{
		currentConfig: u.newConfig,
		currentImage: u.newImage,
	}
}

// This would reverse the old / new configs and images.
// For cases where we might need to roll back to the older version
// (such as when an operation in dn.update() when something fails),
// we can regenerate the onDiskConfig to reflect that thusly:
// update.rollback().toOnDiskConfig().
func (u *updateToApply) rollback() *updateToApply {
	return &updateToApply{
		oldConfig: u.newConfig,
		newConfig: u.oldConfig,
		oldImage: u.newImage,
		newImage: u.oldImage,
	}
}

func (u *updateToApply) toIgnUpdateToApply() (*ignUpdateToApply, error) {
	// Parsing and error handling code omitted for brevity.
}

// This would hold an Ignition representation of the `updateToApply`:
type ignUpdateToApply struct {
	oldIgnConfig ign3types.Config
	newIgnConfig ign3types.Config
}

// Complete with its own rollback method to reverse what is old and new.
func (i *ignUpdateToApply) rollback() *ignUpdateToApply {
	return &ignUpdateToApply{
		oldIgnConfig: u.newIgnConfig,
		newIgnConfig: u.oldIgnConfig,
	}
}

// stateAndConfigs gains a method to create the updateToApply struct.
func (s *stateAndConfigs) toUpdateToApply() *updateToApply {
	return &updateToApply{
		oldConfig: s.currentConfig,
		newConfig: s.desiredConfig,
		oldImage: s.currentImage,
		newImage; s.desiredImage,
	}
}

// updateFromCluster gains a method to create the updateToApply struct.
func (u *updateFromCluster) toUpdateToApply() *updateToApply {
	return &updateToApply{
		oldConfig: u.currentConfig,
		newConfig: u.desiredConfig,
		oldImage: u.currentImage,
		newImage: u.desiredImage,
	}
}
```

We could then modify everything that accepts an old / new -or- current / desired MachineConfig or Ignition pair to consume the `updateToApply` and `ignUpdateToApply` structs. Overall, I think this would reduce confusion and allow some of these transformations and state changes to be more readily testable.

Obviously, this is not something that I will implement in this PR. It is purely for future thought and consideration.